web/client/plugins/dynamicLegend/components/DynamicLegend.jsx

```diff
@@ -0,0 +1,104 @@
+import React from 'react';
+import { keepNode, isLayerVisible } from '../../../selectors/dynamiclegend';
+import { getResolutions } from '../../../utils/MapUtils';
+import Message from '../../../components/I18N/Message';
+import ResizableModal from '../../../components/misc/ResizableModal';
+import { ControlledTOC } from '../../TOC/components/TOC';
+import DefaultGroup from '../../TOC/components/DefaultGroup';
+import DefaultLayer from '../../TOC/components/DefaultLayer';
+
+import '../assets/dynamicLegend.css';
+
+function applyVersionParamToLegend(layer) {
+    // we need to pass a parameter that invalidate the cache for GetLegendGraphic
+    // all layer inside the dataset viewer apply a new _v_ param each time we switch page
+    return { ...layer, legendParams: { ...layer?.legendParams, _v_: layer?._v_ } };
+}
+
+function filterNode(node, currentResolution) {
+    const nodes = Array.isArray(node.nodes) ? node.nodes.filter(keepNode).map(applyVersionParamToLegend).map(n => filterNode(n, currentResolution)) : undefined;
+
+    return {
+        ...node,
+        isVisible: (node.visibility ?? true) && (nodes ? nodes.length > 0 && nodes.some(n => n.isVisible) : isLayerVisible(node, currentResolution)),
+        ...(nodes && { nodes })
+    };
+}
+
+export default ({
+    layers,
+    onUpdateNode,
+    currentZoomLvl,
+    onClose,
+    isVisible,
+    groups,
+    mapBbox
+}) => {
+    const layerDict = layers.reduce((acc, layer) => ({
+        ...acc,
+        ...{
+            [layer.id]: {
+                ...layer,
+                ...{enableDynamicLegend: true, enableInteractiveLegend: false}
+            }
+        }
+    }), {});
+    const getVisibilityStyle = nodeVisibility => ({
+        opacity: nodeVisibility ? 1 : 0,
+        height: nodeVisibility ? "auto" : "0"
+    });
+
+    const customGroupNodeComponent = props => (
+        <div style={getVisibilityStyle(props.node?.isVisible ?? true)}>
+            <DefaultGroup {...props} />
+        </div>
+    );
+    const customLayerNodeComponent = props => {
+        const layer = layerDict[props.node.id];
+        if (!layer) {
+            return null;
+        }
+
+        return (
+            <div style={getVisibilityStyle(props.node?.isVisible ?? true)}>
+                <DefaultLayer {...props} node={layer} />
+            </div>
+        );
+    };

```

Comment:
In https://github.com/allyoucanmap/MapStore2/commit/a11b6736dd0cc5175977ec75a212d3c7d205d309 this component has been changed to avoid too many re-render inside the custom nodes, changes:

- moved customGroupNodeComponent and customLayerNodeComponent outside the render function to avoid re-render on each prop update of DynamicLegend component
- reviewed the hiding logic of node by exposing and using the getNodeStyle prop in ControlledTOC